README.md
<!-- {{{1 -->

    File        : README.md
    Maintainer  : Felix C. Stegerman <flx@obfusk.net>
    Date        : 2019-09-20

    Copyright   : Copyright (C) 2019  Felix C. Stegerman
    Version     : v0.0.1
    License     : GPLv3+, LGPLv3+

<!-- }}}1 -->

<!-- TODO: travis etc. badges -->

[![Build Status](https://travis-ci.org/obfusk/koneko.svg?branch=master)](https://travis-ci.org/obfusk/koneko)
[![GPLv3+](https://img.shields.io/badge/license-GPLv3+-blue.svg)](https://www.gnu.org/licenses/gpl-3.0.html)

## Description

koneko - a concatenative not-quite-lisp for kittens

... TODO ...

### Properties

* concatenative
  - point-free
  - juxtaposition of expressions denotes function composition
* stack-oriented
  - postfix (reverse polish) notation
  - functions consume arguments from the stack
  - functions produce return values on the stack
* lisp-like
  - homoiconic
  - blocks/lambdas (anonymous functions)
  - named arguments/points (lexically scoped)
* functional
  - only immutable data structures
  - does have side effects (I/O)
  - strict evaluation
* dynamically, strongly typed

<!--
  * strict -> streams
  * types -> ???
  * dynamic scope???
  * mark I/O as "dirty" to allow for optimizing code that is known to
    be referentially transparent? vs clojure?
-->

## Examples

### Hello World

```bash
$ koneko -e '"Hello, World!" say'
hello, World!
```

```bash
$ koneko
>>> "Hello, World!" say
Hello, World!
>>> ^D
```

### ...

... TODO ...

### Miscellaneous

NB: WIP

```
>>> , :twice1 [ f . f f ] def             ; with named arguments
>>> 42 [ 1 + ] twice1
44

>>> , :twice2 [ dup 'call dip call ] def  ; points-free
>>> 42 [ 1 + ] twice2
44

>>> , :twice3 [ f . [ f f ] ] def         ; "curried"
>>> 0 [ 1 + ] twice3 twice3 twice3 twice3 call
16

>>> 0 [ 1 + ] [ twice3 twice3 ] twice3 call call
16
>>> 0 [ 1 + ] [ twice3 ] twice3 twice3 call call
16

>>> 0 [ 1 + ] [ twice3 twice3 ] twice3 twice3 call call
256
>>> 0 [ 1 + ] [ twice3 ] twice3 twice3 twice3 call call
256

>>> 0 [ 1 + ] ???                         ; TODO
65536

>>> clear-stack
>>> ( 1 2 3 ) uncons show-stack
( 2 3 )
1
>>> cons
( 1 2 3 )

>>> , :mymap [ f . dup empty? [ ] [ uncons 'f dip 'f mymap cons ] if ] def
>>> [ dup * ] mymap
( 1 4 9 )
```

<!-- [ ] <==> [ drop () ] if empty? -->

## The Language

### Type System

For now, koneko will use dynamic strong typing; an optional static
type system with inference may be added in the future.

### Comments & Whitespace

```
; comments start with a semicolon and end at the end of the line

( 1 2 )   ; tokens are delimited by whitespace
( 3,4 )   ; commas are whitespace
```

<!-- doctest !!! -->

### Primitive Data Types

Nil, Bool, Int, Float, Str, Kwd, Rx

```
>>> nil             ; nothing to see here...
nil

>>> #t              ; true
#t
>>> #f              ; false (nil and #f are falsy, everything else is truthy)
#f

>>> 42              ; integer
42
>>> 0x20
32

>>> 3.14            ; floating point
3.14

>>> "spam & eggs"   ; string
"spam & eggs"
>>> "\u732bs"
"猫s"
>>> "\x20"
" "

>>> :key-word       ; keyword
:key-word
>>> :"keyword that is not a valid identifier"
:"keyword that is not a valid identifier"

>>> /cat|kitten/    ; regex
/cat|kitten/
```

<!--

* Ident
* Raw*
* Range? ...
* Namespace/Scope, Stack, ...

-->

### Pairs, Lists and Dicts

```
>>> :answer 42 =>                 ; key/value pair
:answer 42 =>
>>> dup .key
:answer
>>> .value
42

>>> ( 1 2 :foo 4 )                ; linked list
( 1 2 :foo 4 )
>>> dup len
4
>>> 2 get
:foo

>>> { x: 42, :y 99 1 + => }       ; key/value map
{ x: 42, y: 100 }
>>> dup len
2
>>> :x get
42
>>> { 32 5 +, 76 3 - => }
{ 37 73 => }
```

<!--

* tokens & substack for { ... }, ( ... ) etc.
* linked list vs dynamic array ???

-->

### Records

```
>>> , :Point ( :x :y ) defrecord  ; define record type
>>> Point{ x: 1, y: -1 }          ; create record instance
Point{ x: 1, y: -1 }
>>> dup .x
1
>>> dup .y
-1
>>> { x: 99 } update              ; update record (creates a new one)
Point{ x: 99, y: -1 }
```

<!-- ADTs ??? -->

### Blocks

aka Lambdas

```
>>> , :myblock [ 42 ] def         ; a block that pushes 42 onto the stack
>>> 'myblock                      ; put it on the stack
[ 42 ]
>>> call                          ; call the block on the stack
42
>>> myblock
42

>>> clear-stack
>>> , :myswap [ x y . 'y 'x ] def ; a block with named arguments
>>> 1 2 myswap
1
>>> show-stack
1
2
```

<!--

* arity? args + body
* ..., variable arity, apply, ...
* curry, partial?
* show-stack + reverse order

-->

### Protocols

aka Interfaces

... TODO ...

<!--

* match by partial spec like ocaml!
* !foo vs just foo
* store/match how?

-->

### Sugar

```
>>> answer: 42                              ; pair with keyword as key
:answer 42 =>
>>> :answer 42 =>                           ; desugared
:answer 42 =>

>>> Point{ x: 1, y: 2 }                     ; constructor with "keys"
Point{ x: 1, y: 2 }
>>> { :x 1 =>, :y 2 => } 'Point construct   ; completely desugared
Point{ x: 1, y: 2 }

>>> dup .x                                  ; record field access
1
>>> :x get                                  ; desugared
1

>>> 1 2
2
>>> swap                                    ; push & call
1
>>> 'swap call                              ; explicit push, then call
2

>>> swap( 1 2 )                             ; non-postfix apply
1
>>> ( 1 2 ) 'swap apply                     ; desugared
1
>>> 1 2 swap                                ; same as call (for fixed arity)
1

>>> 1 `+ 2                                  ; shift token (use w/ caution)
3
>>> 1 2 +                                   ; desugared
3
```

```
>>> [[ 1 2 3 + + ]]                         ; grouped expression
6
```

... TODO ...

<!--

* list, dict, quote, grouped expr -> parse tree

* construct does what?
* ( ... ) construct-from-list ??? NO b/c is apply
* Point( 1 2 ) -> Point{ x: 1, y: 2 } ???

* overloaded get for dict, list, record
* !foo, !foo( ... )

* reader vs eval vs primitive vs builtin vs prelude vs stdlib
* [[ ]] is not reader-only sugar

-->

### Quoting

NB: WIP

```
>>> , :x 1 def
>>> , :y 2 def
>>> , :z 3 def

>>> ( 'x 'y 'z )
( 1 2 3 )
>>> '( x y z )
( 1 2 3 )
```

... TODO ...

<!-- macros? -->

### Variables

```
>>> , :answer 42 def            ; define a variable in the current namespace
>>> , :inc [ 1 + ] def          ; can be a value or a block
>>> 'answer inc
43
```

### Primitives

```
>>> 41 [ 1 + ]
[ 1 + ]
>>> call                        ; call the block at the top of the stack
42

>>> 1 2 <
#t
>>> [ :less ] [ :greater ] if   ; conditional
:less
```

... TODO ...

<!-- really __call__, __if__, ... -->

### Builtins

... TODO ...

### Prelude

```
>>> clear-stack
>>> 1 2 show-stack
2
1
>>> ; :swap [ x y . 'y 'x ] def
>>> swap show-stack               ; swap top 2 values on stack
1
2
>>> ; :dup [ x . 'x 'x ] def
>>> dup show-stack                ; dup(licate) top of stack
1
1
2
>>> ; :drop [ x . ] def
>>> drop show-stack               ; drop (remove) top of stack
1
2
>>> ; :dip [ x f . f 'x ] def
>>> 3 '+ dip *                    ; remove x, call f, push x
9
```

... TODO ...

<!-- builtin vs prelude -->

### Standard Library

<!-- lib/ -->

... TODO...

<!-- DSLs, Shell, FFI, ... -->

### ...

... TODO ...

## Help

... TODO ...

## Specs & Docs

... TODO ...

## Requirements

... TODO ...

## Installing

... TODO ...

## TODO

* design
* document
* implement
* ???
* profit!

<!--

* look at notes and old python stuff
* haskell features (Text, not String)
* repl

* dyn vars?
* iterators?

-->

## License

### Interpreter

[GPLv3+](https://www.gnu.org/licenses/gpl-3.0.html).

### Standard Library

(i.e. `lib/*.knk`)

[LGPLv3+](https://www.gnu.org/licenses/lgpl-3.0.html).

<!-- vim: set tw=70 sw=2 sts=2 et fdm=marker : -->


CHANGELOG.md
# Revision history for koneko

## 0.0.1 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


koneko/read.py
# --                                                            ; {{{1
#
# File        : koneko/read.py
# Maintainer  : Felix C. Stegerman <flx@obfusk.net>
# Date        : 2017-11-20
#
# Copyright   : Copyright (C) 2017  Felix C. Stegerman
# Version     : v0.0.1
# License     : GPLv3+
#
# --                                                            ; }}}1

                                                                # {{{1
r"""
... TODO ...

>>> read("")
TODO

... TODO ...
"""                                                             # }}}1

import sys

import re as _re, regex
sys.modules["re"] = regex     # s/regex/re/ for pyparsing *sigh*
import pyparsing as P
sys.modules["re"] = _re

from . import data as D
from . import misc as M

# TODO: .foo, !foo, foo{, ...; no RawDict, List, etc.
def _make_parser():                                             # {{{1
  # NB: the order in which matched are tried is important;
  # e.g. float before int, etc.

  c, g, r, s  = P.Combine, P.Group, P.Regex, P.Suppress
  om, op, zm  = P.OneOrMore, P.Optional, P.ZeroOrMore
  ibody       = r(M.RX_IDENT_BODY).leaveWhitespace()
  lit         = lambda x: r(regex.escape(x))          # escape + rx
  l           = lambda x: s(lit(x)).leaveWhitespace() # literal
  k           = lambda x: lit(x) + ~ibody             # keyword
  n           = lambda x, name: x.setName(name)       # name it

  ws, term    = s(r(M.RX_SPACE)).leaveWhitespace(), P.Forward()
  comm        = n(r(M.RX_COMMENT), "comment")
  reserved    = P.Word(M.S_BRACKETS, exact = 1) + ~ibody
  ident       = ~reserved + r(M.RX_IDENT_C) + ~ibody
  fident      = n(c(zm(c(ident + lit("."))) + ident),
                  "identifier").setParseAction(P.tokenMap(D.Ident))

  hexint      = r(M.RX_HEXINT).setParseAction(P.tokenMap(int, 16))
  binint      = r(M.RX_BININT).setParseAction(P.tokenMap(int, 2))

  nil         = n(k("nil"), "nil").setParseAction(lambda t: [None])
  bool_       = n(k("#t") | k("#f"), "bool") \
                .setParseAction(lambda t: t[0] == "#t")
  float_      = n(P.pyparsing_common.sci_real, "float")   # b4 int!
  int_        = n(hexint | binint |
                  P.pyparsing_common.signed_integer, "int")
  str_        = n(c(l('"') + r(M.RX_STRING) + l('"')), "string") \
                .setParseAction(P.tokenMap(_parse_str))
  kwd         = n(c(l(":") + (ident | str_)), "keyword") \
                .setParseAction(P.tokenMap(D.Kwd))
  regex_      = n(c(l('/"') + r(M.RX_RAWSTRING) + l('"')), "regex") \
                .setParseAction(P.tokenMap(D.Regex))

  list_       = n(g(s(k("(")) + ws + op(om(term + ws)) + s(k(")"))),
                  "list").setParseAction(lambda t: [D.List(t[0])])
  dict_       = n(g(s(k("{")) + ws + op(om((term + ws)*2)) +
                    s(k("}"))), "dict") \
                .setParseAction(lambda t: [D.RawDict(tuple(t[0]))])

  record      = ...

  block       = n(g(s(k("[")) + ws +
                    op(g(om(ident + ws))("params") + s(k(".")) + ws) +
                    op(g(om(term + ws))("code")) +
                    s(k("]"))), "block").setParseAction(_parse_block)

  quote       = n(c(l("'") + fident), "quote") \
                .setParseAction(lambda t: D.Quote(D.Ident(t[0])))
  shift       = n(r("`+").leaveWhitespace() + fident, "shift") \
                .setParseAction(lambda t: D.Shift(t[1], len(t[0])))

  shorthands  = ... # .foo !foo

  value       = nil | bool_ | float_ | int_ | str_ | kwd | \
                regex_ | list_ | dict_ | block | quote          # TODO
  term       << ( value | fident | shift )
  return op(zm(term + ws) + term).ignore(comm)
                                                                # }}}1

# TODO: why is this called twice? _parse_str (etc?) too...
def _parse_block(t):
  params  = tuple(map(D.Ident, t[0].get("params", ())))
  code    = tuple(             t[0].get("code"  , ()) )
  return D.RawBlock(params, code)

def _parse_str(s):                                              # {{{1
  r"""
  Parses valid string contents; handles \uXXXX and \xXX etc.

  >>> s = '"\\u732bs like string"'; s[3:7]
  '732b'
  >>> _parse_str(s[1:-1])
  '猫s like string'
  >>> _parse_str(r"\"\\...")
  '"\\...'
  """

  t, i, n = [], 0, len(s)
  while i < n:
    if s[i] == '\\':
      if s[i+1] in '\\"':
        t.append(s[i+1]); i += 2
      elif s[i+1] == 'x':
        t.append(chr(int(s[i+2:i+4], 16))); i += 4
      elif s[i+1] == 'u':
        t.append(chr(int(s[i+2:i+6], 16))); i += 6
      else: raise RuntimeError("WTF")
    else:
      t.append(s[i]); i += 1
  return "".join(t)
                                                                # }}}1

_parser = _make_parser()

# TODO: more tests, also: failures; repr() -> show()
def parse(s):                                                   # {{{1
                                                                # {{{2
  r"""
  Parse code into an intermediate result.

  >>> parse("nil")[0] is None
  True
  >>> parse("#t")[0]
  True
  >>> parse("#f")[0]
  False
  >>> parse("-4.2")[0]
  -4.2
  >>> parse("0x20")[0]
  32
  >>> s = '"\\u732bs like string"'; s[3:7]
  '732b'
  >>> parse(s)[0]
  '猫s like string'
  >>> parse('"\x20"')[0]
  ' '
  >>> parse(':"foo-\\u732b-bar"')[0]
  :foo-猫-bar
  >>> parse('/"foo\\bar"')[0]
  /"foo\\bar"

  >>> parse("( 1 2 :foo )")[0]
  ( 1 2 :foo )
  >>> parse("{ x 42 'y [ 37 ] }")[0]
  '{ x 42, 'y '[ 37 ] }

  >>> parse("[,]")[0]
  '[ ]
  >>> parse("[ 42 ]")[0]
  '[ 42 ]
  >>> parse("[ x . ]")[0]
  '[ x . ]
  >>> parse("[ x . 42 ]")[0]
  '[ x . 42 ]

  >>> type(parse("'foo")[0]).__name__
  'Quote'
  >>> type(parse("foo")[0]).__name__
  'Ident'
  >>> type(parse("`foo")[0]).__name__
  'Shift'

  >>> parse("'foo")[0]
  'foo
  >>> parse("foo")[0]
  foo
  >>> parse("`foo")[0]
  `foo

  >>> print(" ".join(map(repr, parse(" ,#t 42 'q,"))))
  True 42 'q

  ... TODO ...
  """                                                           # }}}2

  return list(_parser.parseString(s.strip(M.S_SPACE), True))
                                                                # }}}1

# TODO
# * `foo, .foo, !foo, **call**, ...
def read(s):
  """Parse code into an expression."""
  ...

# ... TODO ...

if __name__ == "__main__":
  import doctest
  if doctest.testmod()[0]: sys.exit(1)

# vim: set tw=70 sw=2 sts=2 et fdm=marker :
